src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Query/PropertyPages/LaunchProfileFromProjectDataProducer.cs

```diff
@@ -25,64 +20,12 @@ public LaunchProfileFromProjectDataProducer(ILaunchProfilePropertiesAvailableSta
-            {
-                return createLaunchProfileValues();
-            }
-
-            return Enumerable.Empty<IEntityValue>();
-
-            IEnumerable<IEntityValue> createLaunchProfileValues()
+            if (providerState.Services.ExportProvider.GetExportedValueOrDefault<IProjectLaunchProfileHandler>() is IProjectLaunchProfileHandler launchProfileHandler)
             {
-                Dictionary<string, Rule> debugRules = new();
-                foreach (Rule rule in DebugUtilities.GetDebugChildRules(projectCatalog))
-                {
-                    if (rule.Metadata.TryGetValue("CommandName", out object? commandNameObj)
-                        && commandNameObj is string commandName)
-                    {
-                        debugRules[commandName] = rule;
-                    }
-                }
-
-                if (launchSettings is IVersionedLaunchSettings versionedLaunchSettings)
-                {
-                    queryExecutionContext.ReportInputDataVersion(launchSettingsTracker.VersionKey, versionedLaunchSettings.Version);
-                }
-
-                IProjectState projectState = new LaunchProfileProjectState(project, launchSettingsProvider, launchSettingsTracker);
-
-                foreach ((int index, ProjectSystem.Debug.ILaunchProfile profile) in launchSettings.Profiles.WithIndices())
-                {
-                    if (!Strings.IsNullOrEmpty(profile.Name)
-                        && !Strings.IsNullOrEmpty(profile.CommandName)
-                        && debugRules.TryGetValue(profile.CommandName, out Rule rule))
-                    {
-                        QueryProjectPropertiesContext propertiesContext = new(
-                            isProjectFile: true,
-                            file: project.FullPath,
-                            itemType: LaunchProfileProjectItemProvider.ItemType,
-                            itemName: profile.Name);
-
-                        IEntityValue launchProfileValue = CreateLaunchProfileValue(queryExecutionContext, parent, propertiesContext, rule, index, projectState);
-                        yield return launchProfileValue;
-                    }
-                }
+                return launchProfileHandler.RetrieveAllLaunchProfileEntitiesAsync(queryExecutionContext, parent, _properties);
             }
-        }
-
-        private IEntityValue CreateLaunchProfileValue(IQueryExecutionContext queryExecutionContext, IEntityValue parent, QueryProjectPropertiesContext propertiesContext, Rule rule, int order, IProjectState projectState)
-        {
-            EntityIdentity identity = LaunchProfileDataProducer.CreateLaunchProfileId(parent, propertiesContext.ItemType!, propertiesContext.ItemName!);
 
-            return LaunchProfileDataProducer.CreateLaunchProfileValue(queryExecutionContext, identity, propertiesContext, rule, order, projectState, _properties);
+            return Task.FromResult(Enumerable.Empty<IEntityValue>());
```

Comment:
This allocation can be avoided via the change in PR #7482. @tmeschter if you want to make that change here while I'm offline, feel free to merge that PR.

Reply:
I'll pick that up in my next change.